File: Labo1 - java - template/Labo1 - java/app/src/main/java/ch/heigvd/iict/sym/labo1/MainActivity.java
package ch.heigvd.iict.sym.labo1;

import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.EditText;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.util.Pair;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    // on définit une liste de couples e-mail / mot de passe
    // ceci est fait juste pour simplifier ce premier laboratoire,
    // mais il est évident que de hardcoder ceux-ci est une pratique à éviter à tout prix...
    private List<Pair<String, String>> credentials = new ArrayList<>(Arrays.asList(
                                                        new Pair<>("user1@heig-vd.ch","1234"),
                                                        new Pair<>("user2@heig-vd.ch","abcd")));

    private EditText email;
    private EditText password;
    private Button cancelButton;
    private Button validateButton;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        // l'appel à la méthode onCreate de la super classe est obligatoire
        super.onCreate(savedInstanceState);
        // on définit le layout à utiliser pour l'affichage
        setContentView(R.layout.activity_main);

        // on va maintenant lier le code avec les éléments graphiques (champs texts, boutons, etc.)
        // présents dans le layout (nous allons utiliser l'id défini dans le layout, le cast est
        // réalisé automatiquement)
        email = findViewById(R.id.main_email);
        password = findViewById(R.id.main_password);
        cancelButton = findViewById(R.id.main_cancel);
        validateButton = findViewById(R.id.main_validate);

        //mise en place des événements
        cancelButton.setOnClickListener(view -> {
            //on va vider les champs de la page de login lors du clique sur le bouton Cancel
            email.getText().clear();
            password.getText().clear();
            // on annule les éventuels messages d'erreur présents sur les champs de saisie
            email.setError(null);
            password.setError(null);
        });

        validateButton.setOnClickListener(view -> {
            //on réinitialise les messages d'erreur
            email.setError(null);
            password.setError(null);

            //on récupère le contenu de deux champs dans des variables de type String
            String emailInput = email.getText().toString();
            String passwordInput = password.getText().toString();

            if (emailInput.isEmpty() || passwordInput.isEmpty()){
                // on affiche un message dans les logs de l'application
                Log.d(TAG, "Au moins un des deux champs est vide");
                // on affiche un message d'erreur sur les champs qui n'ont pas été renseignés
                // la méthode getString permet de charger un String depuis les ressources de
                // l'application à partir de son id
                if (emailInput.isEmpty())
                    email.setError(getString(R.string.main_mandatory_field));
                if (passwordInput.isEmpty())
                    password.setError(getString(R.string.main_mandatory_field));

                return;
            }

            //TODO à compléter...
        });

    }
}

File: Labo1 - kotlin - template/Labo1 - kotlin/app/src/main/java/ch/heigvd/iict/sym/labo1/MainActivity.kt
package ch.heigvd.iict.sym.labo1

import android.app.Activity
import android.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.EditText
import android.content.DialogInterface
import android.content.Intent
import android.provider.AlarmClock.EXTRA_MESSAGE
import android.net.Uri
import android.view.View
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultCallback
import androidx.activity.result.contract.ActivityResultContract
import androidx.activity.result.contract.ActivityResultContracts
import ch.heigvd.iict.sym.labo1.utils.Utils
import java.net.URI


class MainActivity : GenericActivity() {

    // on définit une liste de couples e-mail / mot de passe
    // ceci est fait juste pour simplifier ce premier laboratoire,
    // mais il est évident que de hardcoder ceux-ci est une pratique à éviter à tout prix...
    // /!\ listOf() retourne une List<T> qui est immuable
    private var credentials = listOf(
                                Pair("user1@heig-vd.ch","1234"),
                                Pair("user2@heig-vd.ch","abcd")
                            )

    // le modifieur lateinit permet de définir des variables avec un type non-null
    // sans pour autant les initialiser immédiatement
    private lateinit var email: EditText
    private lateinit var password: EditText
    private lateinit var cancelButton: Button
    private lateinit var validateButton: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        // l'appel à la méthode onCreate de la super classe est obligatoire
        super.onCreate(savedInstanceState)
        // on définit le layout à utiliser pour l'affichage
        setContentView(R.layout.activity_main)

        // on va maintenant lier le code avec les éléments graphiques (champs texts, boutons, etc.)
        // présents dans le layout (nous allons utiliser l'id défini dans le layout, le cast est
        // réalisé automatiquement)
        email = findViewById(R.id.main_email)
        password = findViewById(R.id.main_password)
        cancelButton = findViewById(R.id.main_cancel)
        validateButton = findViewById(R.id.main_validate)
        // Kotlin, au travers des Android Kotlin Extensions permet d'automatiser encore plus cette
        // étape en créant automatiquement les variables pour tous les éléments graphiques présents
        // dans le layout et disposant d'un id
        // cf. https://medium.com/@temidjoy/findviewbyid-vs-android-kotlin-extensions-7db3c6cc1d0a

        //mise en place des événements
        cancelButton.setOnClickListener {
            //on va vider les champs de la page de login lors du clique sur le bouton Cancel
            email.text?.clear()
            password.text?.clear()
            // on annule les éventuels messages d'erreur présents sur les champs de saisie
            email.error = null
            password.error = null
        }

        validateButton.setOnClickListener { it ->
            //on réinitialise les messages d'erreur
            email.error = null
            password.error = null

            //on récupère le contenu de deux champs dans des variables de type String
            val emailInput = email.text?.toString()
            val passwordInput = password.text?.toString()
            val credentialsInput = Pair(emailInput,passwordInput)

            if(emailInput.isNullOrEmpty() or passwordInput.isNullOrEmpty()) {
                // on affiche un message dans les logs de l'application
                Log.d(TAG, "Au moins un des deux champs est vide")
                // on affiche un message d'erreur sur les champs qui n'ont pas été renseignés
                // la méthode getString permet de charger un String depuis les ressources de
                // l'application à partir de son id
                if(emailInput.isNullOrEmpty())
                    email.error = getString(R.string.main_mandatory_field)
                if(passwordInput.isNullOrEmpty())
                    password.error = getString(R.string.main_mandatory_field)



                // Pour les fonctions lambda, on doit préciser à quelle fonction l'appel à return
                // doit être appliqué
                return@setOnClickListener
            }
            //Etape 3.1
            else if(!Utils.checkEmailFormat(emailInput)/*!emailInput!!.contains('@')*/) {
                Utils.sendEmailError(applicationContext);
            }
            //Etape 4.1
            else if(credentials.contains(credentialsInput)) {
                //TODO lancer nouvelle activité
                val intent = Intent(this, ProfileActivity::class.java).apply {
                    putExtra(EXTRA_MESSAGE, emailInput)
                }
                startActivity(intent)
            }
            //Etape 3.2
            else {
                val builder1: AlertDialog.Builder = AlertDialog.Builder(this)
                builder1.setMessage("The credentials entered are not valid")
                builder1.setCancelable(true)

                builder1.setPositiveButton(
                    "Ok",
                    DialogInterface.OnClickListener { dialog, id -> dialog.cancel() })

                val alert11: AlertDialog = builder1.create()
                alert11.show()
            }

            //TODO à compléter...

        }
    }

    val getContent = registerForActivityResult(ActivityResultContracts.StartActivityForResult())
    {
            result: ActivityResult ->
        if (result.resultCode == Activity.RESULT_OK) {
            val email = result.data?.getStringExtra("newEmail");
            val password = result.data?.getStringExtra("newPassword").toString();

            if (!email.isNullOrEmpty() && !password.isNullOrEmpty()){
                credentials = credentials + Pair(email, password);
            }
        }
    }

    fun sendMessage(view : View) {
        val intent = Intent(this, Register::class.java);
        getContent.launch(intent);
    }

    // En Kotlin, les variables static ne sont pas tout à fait comme en Java
    // pour des raison de lisibilité du code, les variables et méthodes static
    // d'une classe doivent être regroupées dans un bloc à part: le companion object
    // cela permet d'avoir un aperçu plus rapide de tous les éléments static d'une classe
    // sans devoir trouver le modifieur dans la définition de ceux-ci, qui peuvent être mélangé
    // avec les autres éléments non-static de la classe
    companion object {
        private const val TAG: String = "MainActivity"
    }

}
